### ISSUE TITLE
Inference with gemma-reranker

### ISSUE THREAD

**Issue by Rispo997** [FlagOpen/FlagEmbedding#702]
Hello guys, thank you for your framework! Results seems promising and I am trying to finetune the gemma-reranker, but I have a few questions.
1) The fine-tune toy data example for gemma contains also the prompt. Is the prompt necessary? I would be interested only in the re-ranking function, so I don't need any prompt. Can I still train without, or do I need to create one ad-hoc? Did you use any other prompt during training other than the one present in the example (I would prefer to get a score rather than a Yes or Not).
2) I have fine-tuned the model, then merged the lora weights with the base model and I tried to do inference.
If I do inference with your framework (by calling reranked.compute_score([query, document]) I always get NaN as scores. Might the reason be that I gave a prompt during training whereas I don't give any to the compute_score? Did you ever observe this behaviour?  
3) Inference through your huggingface example seem to not give NaN, but I don't fully understand how you compute the scores:
`yes_loc = tokenizer('Yes', add_special_tokens=False)['input_ids'][0]`
So your score becomes the logit associated to the 'Yes' token? And you ignore the logits of the 'No' token? Why is this computed differently inside the compute_scores function?

**Comment by 545999961** [comment]
1. Adding a prompt can improve performance during fine-tuning and inference. The output score is the logits for 'Yes'. It's recommended to create a prompt during fine-tuning. We trained different datasets with different prompts; the prompt in the example is specifically for the QA dataset.
2. I have ever observed this behavior when using fp-16 or setting a large learning rage, but if you fine-tune with a prompt, it's necessary to perform inference with the prompt.
3. The logits for 'No' are being ignored. We use the logits for 'Yes' to represent the score, the `compute_score` function is the same as the example.

**Comment by Rispo997** [comment]
Thank you @545999961 now it's clear to me.
My only doubt remain about the differences between the huggingface example and the compute_score().
With the former I get a score, whereas the compute_score() function always gives me nan, even with the same prompt.
I don't think it's something related to the training process, otherwise I'd get nan in both cases. Does the huggingface example and your compute_score() give the exact same output to you ?

**Comment by 545999961** [comment]
I run the code but have the same result：
transformers:
```
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

def get_inputs(pairs, tokenizer, prompt=None, max_length=1024):
    if prompt is None:
        prompt = "Given a query A and a passage B, determine whether the passage contains an answer to the query by providing a prediction of either 'Yes' or 'No'."
    sep = "\n"
    prompt_inputs = tokenizer(prompt,
                              return_tensors=None,
                              add_special_tokens=False)['input_ids']
    sep_inputs = tokenizer(sep,
                           return_tensors=None,
                           add_special_tokens=False)['input_ids']
    inputs = []
    for query, passage in pairs:
        query_inputs = tokenizer(f'A: {query}',
                                 return_tensors=None,
                                 add_special_tokens=False,
                                 max_length=max_length * 3 // 4,
                                 truncation=True)
        passage_inputs = tokenizer(f'B: {passage}',
                                   return_tensors=None,
                                   add_special_tokens=False,
                                   max_length=max_length,
                                   truncation=True)
        item = tokenizer.prepare_for_model(
            [tokenizer.bos_token_id] + query_inputs['input_ids'],
            sep_inputs + passage_inputs['input_ids'],
            truncation='only_second',
            max_length=max_length,
            padding=False,
            return_attention_mask=False,
            return_token_type_ids=False,
            add_special_tokens=False
        )
        item['input_ids'] = item['input_ids'] + sep_inputs + prompt_inputs
        item['attention_mask'] = [1] * len(item['input_ids'])
        inputs.append(item)
    return tokenizer.pad(
            inputs,
            padding=True,
            max_length=max_length + len(sep_inputs) + len(prompt_inputs),
            pad_to_multiple_of=8,
            return_tensors='pt',
    )

tokenizer = AutoTokenizer.from_pretrained('BAAI/bge-reranker-v2-gemma')
model = AutoModelForCausalLM.from_pretrained('BAAI/bge-reranker-v2-gemma')
yes_loc = tokenizer('Yes', add_special_tokens=False)['input_ids'][0]
model.eval()

pairs = [['what is panda?', 'hi'], ['what is panda?', 'The giant panda (Ailuropoda melanoleuca), sometimes called a panda bear or simply panda, is a bear species endemic to China.']]
with torch.no_grad():
    inputs = get_inputs(pairs, tokenizer)
    scores = model(**inputs, return_dict=True).logits[:, -1, yes_loc].view(-1, ).float()
    print(scores)
```
FlagEmbedding:
```
from FlagEmbedding import FlagLLMReranker
reranker = FlagLLMReranker('BAAI/bge-reranker-v2-gemma')

scores = reranker.compute_score([['what is panda?', 'hi'], ['what is panda?', 'The giant panda (Ailuropoda melanoleuca), sometimes called a panda bear or simply panda, is a bear species endemic to China.']])
print(scores)
```

**Comment by Rispo997** [comment]
@545999961 Thank you, you were right. The problem was that I set fp16 when loading the model.
